### ISSUE TITLE
Support Variables in arange etc.

### ISSUE THREAD

**Issue by jl-wynen** [scipp/scipp#2140]
Building an arange from Variables with units is quite tricky: (The example computes a starting point in addition, that is not the point of this issue. But this is a real world example.)
```.py
def calculate_pulse_times(dim, tmin, tmax, period, *, unit=None):
    unit = period.unit if unit is None else unit
    n_steps = sc.to_unit(tmin / period, sc.units.one, copy=False)
    first_time = period * math.ceil(n_steps.value)
    return sc.arange(dim, 
                     sc.to_unit(first_time, unit).value,
                     sc.to_unit(tmax, unit).value,
                     sc.to_unit(period, unit).value,
                     unit=unit)
```
This example is probably missing a lot of edge cases. It breaks if the unit of `period` is smaller than that of `tmin` and `period` as integer dtype as that causes `sc.to_unit(period, unit).value` to be zero.

I think that `arange` (and company) should accept scalar Variables as arguments to encapsulate the unit handling.

**Comment by SimonHeybrock** [comment]
Related: What is also frequently used is to "split" an existing interval. For example, we may have binned data with only a single bin, but want N bins instead. Something like
```python
limits = sc.array(dim=dim, unit=unit, values=[100.0, 200.0])
sc.linspace(limits=limits, num=100)   # dim can also be deduced from limits
```
might make sense?

**Comment by jl-wynen** [comment]
I think in this case it makes more sense to add a function with a different name. I'd rather avoid mutually exclusive sets of arguments like `limits` and (`start`, `stop`, etc.). Maybe `linspace_within(limits)` or `derive_linspace` (the latter could be a method, though that is maybe not a great idea.)

**Comment by SimonHeybrock** [comment]
Yes, a different name would probably be the correct choice.

**Comment by SimonHeybrock** [comment]
We have previously struggled with functions that supported a set of mutually exclusive arguments or argument types. In this case, if start/stop/step are variables then `unit` must not be given. This yields overhead in docstrings and the wrapper, which needs to check for this. Can we find a solution that avoids this?

I am not sure I am suggesting any of these, but here are some options:
- *Only* support variables in `arange`. User has to use `sc.arange(sc.scalar(start, unit), sc.scalar(step, unit), sc.scalar(step, unit))` which is very cumbersome.
- Use a differently named function for the two variants. I have no intuitive suggestion.
- Add a new mechanism entirely, i.e., when start/stop are variables, don't use `arange`, but something like:
  ```python
  sc.interval_to_range(sc.concat(start, stop, dim), step=step)
  sc.interval_to_linspace(sc.concat(start, stop, dim), num=num)
  # or possibly more concise:
  sc.interval(start, stop, dim).range(step=step)
  sc.interval(start, stop, dim).linspace(step=step)
  ```
  Where in the last option `sc.interval` returns a helper object.

**Comment by jl-wynen** [comment]
My thoughts on your not suggested options:
1. This hurts usability
2. Could work, but potentially hard to remember which is which.
3. The first 2 are quite verbose and one needs to remember two functions. In the other two, what is `interval`? Is it like `slice`? If it has no other uses, we should not add it just for this use case.

Concerning the original suggestion, I don't think that `unit` is mutually exclusive with passing variables. In fact, it might be required or at least useful when `start` and `stop` have different units. It would effectively be like `to_unit` in that it can only perform a restricted set of conversions.

**Comment by SimonHeybrock** [comment]
>     3\. In the other two, what is `interval`? Is it like `slice`?

It is a helper object that you create temporarily, a bit like what you get from `groupby`, but instead of giving you "groups" that you can apply something to it gives you an "interval" that you can manipulate.

>  If it has no other uses, we should not add it just for this use case.

I would not necessarily say so: This is an important and frequently arising situation (creating a new coord is extremely common). `sc.interval` could be the tool that allows splitting this process into two steps:
1. Define the interval.
2. Define a range based on the interval.

**Comment by SimonHeybrock** [comment]
> I don't think that `unit` is mutually exclusive with passing variables. In fact, it might be required or at least useful when `start` and `stop` have different units. It would effectively be like `to_unit` in that it can only perform a restricted set of conversions.

I think we would be getting into dangerous territory here very quickly. If we did this `arange` would end up as a powerful and complex "super function", rather than a simple tool. I feel this would go against the scipp approach of providing small and modular building blocks that do one thing well.

**Comment by SimonHeybrock** [comment]
Related to the proposal in #2508: The solution is partial, since it does not support the case of taking an existing interval and subdividing it. Say we have existing bin bounds, given by a variable of length 2. We would need to:

```python
# probably linspace would be more suitable in this case
edges = sc.arange(coord.dim, coord[0], coord[-1], (coord[-1]-coord[0])/100)
# or
edges = sc.arange(coord.dim, coord[0], coord[-1], 2 * sc.Unit('mm'), unit=coord.unit)
```
While better than what we have now, I wonder if we can't find something that solves this at the same time?

**Comment by jl-wynen** [comment]
I think that would require an API long the lines of
```.py
sc.foo(limits, step)
sc.foo(sc.concat([start, stop], dim), step)
```
`concat` could be moved into `foo`. But then the user needs to specify `dim` as an arg to `foo` which hurts the first case.

**Comment by SimonHeybrock** [comment]
Where `foo` is `arange`, `linspace`, or `geomspace`?

> concat could be moved into foo.

You mean as a convenience to avoid writing `concat` in cases where we do not have an existing interval?

Another subtlety is that `step` should be given as a variable, I suppose? That is, we would not support providing a plain value in this case?

**Comment by jl-wynen** [comment]
> Where foo is arange, linspace, or geomspace?

Or some other function that is more specialised for handling variables as inputs.

> You mean as a convenience to avoid writing concat in cases where we do not have an existing interval?
> 
> Another subtlety is that step should be given as a variable, I suppose? That is, we would not support providing a plain value in this case?

Yes to both

**Comment by jl-wynen** [comment]
I am wondering what is the value of the function that you propose, @SimonHeybrock ?
- Does it only allow 1d length-2 variables as inputs? That would seem oddly specific. But anything else would make the function very complicated.
- How often does this situation arise? And where does the length-2 variable come from?
- Is `sc.arange(coord, (coord[-1]-coord[0])/100)` really that much better than `sc.arange(coord.dim, coord[0], coord[-1], (coord[-1]-coord[0])/100)`?

**Comment by SimonHeybrock** [comment]
Implemented in #2508.
